### PR TITLE
Fixes type annotations for named tensors #27846

### PR DESCRIPTION
**Pull request by physicsrob**
This enables type checking for named tensors, and fixes the underlying problems. 

The bulk of the fix is modifying `gen_pyi.py` to generate reasonable types in `torch/__init__.pyi`.  I took two approaches:  First, I tried to take a generic approach and added `DimnameList` to the magic list of variable argument lists.  Unfortunately that was insufficient for many of the method signatures, so I also added manual definitions for `rename`, `refine_names`, and `unflatten` in `__init__.pyi.in`.

Finally there were a few problems in the doctests that had to be cleaned up so that `test/test_type_hints.py` will run successfully.

### CODE REVIEW

**Comment by physicsrob** [comment]
cc @zou3519 (author of Named Tensors)
cc @t-vi (author of the pyi module this modifies)

File: tools/pyi/gen_pyi.py [hunk 1]
```diff
@@ -300,22 +303,27 @@ def generate_type_hints(fname, decls, is_tensor=False):
...
+                               decl['arguments'][vararg_pos]['dynamic_type'] in {'IntArrayRef', 'TensorList', 'DimnameList'} and
                                (numargs == vararg_pos + 1 or python_args[vararg_pos + 1] == '*') and
                                (not is_tensor or decl['arguments'][0]['name'] == 'self'))
 
         type_hints.append(type_hint)
 
         if have_vararg_version:
             # Two things come into play here: PyTorch has the "magic" that if the first and only positional argument
-            # is an IntArrayRef or TensorList, it will be used as a vararg variant.
+            # is an IntArrayRef, TensorList, or DimnameList, it will be used as a vararg variant.
```
Comment (zou3519) [inline tools/pyi/gen_pyi.py:314]: I'm not sure what is going on here. If I am understanding this correctly, this is only used to generate type annotations for `tensor.align_to`. If that is the case, could we manually bind it instead?

File: test/test_type_hints.py [hunk 2]
```diff
@@ -63,12 +63,6 @@ def get_all_examples():
     """
     blacklist = {
         "_np",
-        "refine_names",
-        "rename",
-        "names",
-        "align_as",
-        "align_to",
-        "unflatten",
```
Comment (zou3519) [inline test/test_type_hints.py:71]: To summarize:
- names just works
- align_as just works
- This PR adds manual entries for unflatten, refine_names, and rename 
- `align_to` works from changing the annotations generation code

Is that an accurate summary of what is going on in this PR?

**Review by zou3519 — COMMENTED** [review]
Some initial questions

File: tools/pyi/gen_pyi.py [hunk 3]
```diff
@@ -135,8 +138,8 @@ def type_to_python(typename, size=None):
         'void*': '_int',    # data_ptr
         'void': 'None',
         'std::string': 'str',
-        'Dimname': 'Union[str, None]',
-        'DimnameList': 'List[Union[str, None]]',
+        'Dimname': 'Union[str, ellipsis]',
```
[the remark this answers is not in the document]
Comment (physicsrob) [reply tools/pyi/gen_pyi.py:141]: I handled this by making the manual type for rename to be `Union[str, None]`.  I was trying to handle the general case here, but if in general Dimname can be None then I agree it should be added back.

File: tools/pyi/gen_pyi.py (hunk 1 above)
Comment (physicsrob) [reply tools/pyi/gen_pyi.py:314]: I started off trying to handle the generic case, and then realized that almost every method required a manual binding.  If more variable argument methods are added to NamedTensor in the future this code should result in automatic bindings, so there is some utility in leaving it in.  Let me know if you think that's unlikely to happen and I'm happy to strip out the changes to the generated and add the manual case for `align_to`.

File: test/test_type_hints.py (hunk 2 above)
Comment (physicsrob) [reply test/test_type_hints.py:71]: Your summary is right on.

I'll wait for feedback on what you'd like to do with the annotation generation code with regard to `align_to`.  I'm happy to strip-out the changes to the generator and add the manual case for `align_to` if you think that approach is preferable.

File: tools/pyi/gen_pyi.py (hunk 1 above)
Comment (zou3519) [reply tools/pyi/gen_pyi.py:314]: I don't see a large number of variable argument methods being added in the near future. The manual case makes it more obvious what is going on and is not too much effort to write for each new op.

File: tools/pyi/gen_pyi.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zou3519) [reply tools/pyi/gen_pyi.py:141]: In general Dimname can be None. Though there's a question if we should represent Dimname as only being strings and then using `Optional[Dimname]` for the operators that do take in None for Dimname.

**Comment by physicsrob** [comment]
Update to address @zou3519 suggestion (manually creating the type stub for `align_to`, remove the automatic logic in `gen_pyi`) and rebased on master.

**Review by zou3519 — APPROVED** [review]
thank you!

**Comment by zou3519** [comment]
clang-format test failure looks unrelated because none of the code in this PR is actually in C++.

**Pull request closed without merging**